core/matrix/fbcsr.cpp

```diff
@@ -0,0 +1,689 @@
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/matrix/fbcsr.hpp>
+
+
+#include <limits>
+#include <map>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/matrix/coo.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/matrix/ell.hpp>
+#include <ginkgo/core/matrix/identity.hpp>
+#include <ginkgo/core/matrix/sellp.hpp>
+#include <ginkgo/core/matrix/sparsity_csr.hpp>
+
+
+#include "core/components/absolute_array.hpp"
+#include "core/components/fill_array.hpp"
+#include "core/components/fixed_block.hpp"
+#include "core/matrix/fbcsr_kernels.hpp"
+
+
+namespace gko {
+namespace matrix {
+namespace fbcsr {
+
+
+GKO_REGISTER_OPERATION(spmv, fbcsr::spmv);
+GKO_REGISTER_OPERATION(advanced_spmv, fbcsr::advanced_spmv);
+GKO_REGISTER_OPERATION(spgemm, fbcsr::spgemm);
+GKO_REGISTER_OPERATION(advanced_spgemm, fbcsr::advanced_spgemm);
+GKO_REGISTER_OPERATION(spgeam, fbcsr::spgeam);
+GKO_REGISTER_OPERATION(convert_to_coo, fbcsr::convert_to_coo);
+GKO_REGISTER_OPERATION(convert_to_csr, fbcsr::convert_to_csr);
+GKO_REGISTER_OPERATION(convert_to_dense, fbcsr::convert_to_dense);
+GKO_REGISTER_OPERATION(calculate_total_cols, fbcsr::calculate_total_cols);
+GKO_REGISTER_OPERATION(transpose, fbcsr::transpose);
+GKO_REGISTER_OPERATION(conj_transpose, fbcsr::conj_transpose);
+GKO_REGISTER_OPERATION(row_permute, fbcsr::row_permute);
+GKO_REGISTER_OPERATION(inverse_row_permute, fbcsr::inverse_row_permute);
+GKO_REGISTER_OPERATION(inverse_column_permute, fbcsr::inverse_column_permute);
+GKO_REGISTER_OPERATION(calculate_max_nnz_per_row,
+                       fbcsr::calculate_max_nnz_per_row);
+GKO_REGISTER_OPERATION(calculate_nonzeros_per_row,
+                       fbcsr::calculate_nonzeros_per_row);
```

Comment:
You implemented these but I don't think they are used?

Reply:
I think you're right. But they're simple and `calculate_nonzeros_per_row` has a test.